### PR TITLE
Update Discord RPC ranking value depending on the ruleset

### LINKED ISSUE #12898: Discord rich presence should show current game mode rank
As of now, the discord rich presence shows current game mode, what the user is doing, and his rank
image

The rank shown however, does not reflect the current game mode that is being played, but it shows only osu!std
https://github.com/ppy/osu/blob/51e52fffd86347dbdb16c1420d1b3df13debfb1b/osu.Desktop/DiscordRichPresence.cs#L108

Lazer version: 2021.515.0

### PR DESCRIPTION
**Pull request by tornac1234**
- Closes https://github.com/ppy/osu/issues/12898
- [x] Update shown ranking depending on ruleset
~Show beatmap star rating~ reverted and will be discussed in #18499

### CODE REVIEW

File: osu.Desktop/DiscordRichPresence.cs [hunk 1]
```diff
@@ -115,6 +125,47 @@ private void updateStatus()
             client.SetPresence(presence);
         }
 
+        private void requestUserRulesetsRankings()
+        {
+            RulesetInfo cachedRuleset = ruleset.Value;
+
+            // When first logging in, we can't send a request with an empty id
```
Comment (bdach) [inline osu.Desktop/DiscordRichPresence.cs:132]: I don't understand. Can you explain why this is done this way? Ideally these two branches should not exist.

File: osu.Desktop/DiscordRichPresence.cs [hunk 2]
```diff
@@ -66,7 +71,7 @@ private void load(IAPIProvider provider, OsuConfigManager config)
                 activity.BindTo(u.NewValue.Activity);
             }, true);
 
-            ruleset.BindValueChanged(_ => updateStatus());
+            ruleset.BindValueChanged(_ => requestUserRulesetsRankings());
```
Comment (bdach) [inline osu.Desktop/DiscordRichPresence.cs:74]: Not sure this should be firing requests every time the ruleset changes.

File: osu.Desktop/DiscordRichPresence.cs [hunk 3]
```diff
@@ -141,7 +192,7 @@ private string getDetails(UserActivity activity)
             switch (activity)
             {
                 case UserActivity.InGame game:
-                    return game.BeatmapInfo.ToString();
+                    return $"[{Math.Round(game.BeatmapInfo.StarRating, 2)}*] {game.BeatmapInfo}";
```
Comment (bdach) [inline osu.Desktop/DiscordRichPresence.cs:195]: Please do not add unrelated changes to diffs like this. This change has nothing to do with the PR's title and wasn't even properly discussed as an addition before.

**Comment by bdach** [comment]
Also not sure if you've read the [relevant issue thread](https://github.com/ppy/osu/issues/12898#issuecomment-846404392) but the rank fetching should be ideally extracted to a common component, because it's data that is useful in more places than just this one functionality.

File: osu.Desktop/DiscordRichPresence.cs (hunk 2 above)
Comment (tornac1234) [reply osu.Desktop/DiscordRichPresence.cs:74]: Is there another way we could get an updated version of the rankings, or a better timing to do it ?

File: osu.Desktop/DiscordRichPresence.cs (hunk 1 above)
Comment (tornac1234) [reply osu.Desktop/DiscordRichPresence.cs:132]: That's my bad I made it because it was fired once at the very beginning when the user hadn't even logged in, but now that the api call is made only once the RPC has initialized, I can remove the part where user ID is not defined

File: osu.Desktop/DiscordRichPresence.cs (hunk 3 above)
Comment (tornac1234) [reply osu.Desktop/DiscordRichPresence.cs:195]: Oh ok, do you think I should update the PR's name to include this change as well or do you think that it's not a relevant change at all ?

Comment (bdach) [reply osu.Desktop/DiscordRichPresence.cs:195]: Would revert for now and re-raise in a separate discussion thread.

**Comment by tornac1234** [comment]
> Also not sure if you've read the [relevant issue thread](https://github.com/ppy/osu/issues/12898#issuecomment-846404392) but the rank fetching should be ideally extracted to a common component, because it's data that is useful in more places than just this one functionality.

Yeah I read it and concluded that a simple way of getting the user's statistics would be to use `GetUsersRequest`. As I'm not really familiar to the project structure yet, I didn't know a better way of doing the thing.

File: osu.Desktop/DiscordRichPresence.cs (hunk 2 above)
Comment (tornac1234) [reply osu.Desktop/DiscordRichPresence.cs:74]: Oh ok, so it would be fine to only make that request part of the "initial" connection request ?

Comment (bdach) [reply osu.Desktop/DiscordRichPresence.cs:74]: I think so but it may be best to wait for a second opinion here.

File: osu.Desktop/DiscordRichPresence.cs (hunk 1 above)
Comment (bdach) [reply osu.Desktop/DiscordRichPresence.cs:132]: If one of the branches is no longer necessary then please remove.

File: osu.Desktop/DiscordRichPresence.cs (hunk 2 above)
Comment (tornac1234) [reply osu.Desktop/DiscordRichPresence.cs:74]: Also i'm not sure where to find that first connection request, is there a specific place/event that could be reach by DiscordRichPresence to know once the client has logged in ?

Comment (bdach) [reply osu.Desktop/DiscordRichPresence.cs]: You'll probably want to fire the request [on change to `LocalUser`](https://github.com/ppy/osu/blob/80fa90f65d149d806a05a2e616f047616591ae04/osu.Desktop/DiscordRichPresence.cs#L65-L72). Which, by the way, is more correct anyway, because with the way you have things the rank display could break if you logged out and then in to a different account for whatever reason.

File: osu.Desktop/DiscordRichPresence.cs [hunk 4]
```diff
@@ -30,6 +32,9 @@ internal class DiscordRichPresence : Component
 
         private IBindable<APIUser> user;
 
+        [Resolved]
+        private IAPIProvider api { get; set; }
```
Comment (bdach) [inline osu.Desktop/DiscordRichPresence.cs:36]: Also just noticed that this is resolved both to a field here and in to a parameter in `load()` (named `provider`) therein. Would be preferred to rename this field to `provider` and remove the `load()` parameter to unify usage.

**Comment by peppy** [comment]
I can't get behind this change. You've added one extra API request to a call which is already being made in [at least one other location](https://github.com/ppy/osu/blob/42e07b7308c05bd0792ceee0ae691a54f74f8e0d/osu.Game/Beatmaps/DifficultyRecommender.cs#L89-L95) with similar context (current ruleset).

Given that there are multiple usages popping up (and more will likely appear in the future, ie. when we want to show the user's current rank / statistics on the user panel) this lookup and fetch process should be moved out of these individual components into a high level.

Potentially handled by `APIAccess` itself, as that is where the initial population currently happens.

It's going to require a bit more work and thinking, but I think it's better than spamming API requests in individual components like this, as it will require cleaning these up in the future (and potentially the code will be forgotten about until we hit some API limits).

**Review by peppy — CHANGES_REQUESTED** [review]
As proposed

**Comment by tornac1234** [comment]
I understood what you said about moving the instructions, but I'm still wondering if I understood it the correct way. Should my modifications be:
1. only client-side : I simply make a call to ``GetUsersRequest`` just after the initial ``GetUserRequest`` has succeeded (because we don't have the ``LocalUser``'s id beforehands)
2. on server-side : create a new type of requests to ressemble the ``GetUsersRequest`` (would be way harder for this "simple" purpose) and be called as ``GetFriendsRequest``

**Comment by tornac1234** [comment]
Something that blocks me right now is that ``DiscordRichPresence``'s ``LocalUser.BindValueChanged`` only works one time (excluding the immediate run). As soon as LocalUser has been modified one time, the callback won't work again. Do you know how I should deal with this issue ?

**Comment by tornac1234** [comment]
> > As soon as `LocalUser` has been modified one time, the callback won't work again
> 
> are you sure the local user is actually changing? the local user changes on logout (to a guest user) and on login, after a succcessful user request (to the actual logged in user). i just tested on `master` and it works as expected there.

I've added another request right after ``GetUserRequest`` and when it finishes, it sets ``localUser.Value``, but the callback doesn't trigger

**Comment by bdach** [comment]
i'm honestly not even sure what you're trying to do at this point so i can't help you. i can't make any sense out of that last comment.

**Comment by tornac1234** [comment]
I'm expecting the BindValueChanged event to be triggered (the callback which is set in ``DiscordRichPresence.load()``) each time ``APIAccess.localUser.Value`` is set. But the event is apparently only triggered the first time that ``APIAccess.localUser.Value`` is set. Though, I'd like to have this event triggered every time ``APIAccess.localUser.Value`` is set

**Comment by tornac1234** [comment]
Ok so this is what it would *roughly* look like, I'm not sure about the order of the requests though

**Comment by LittleEndu** [comment]
The changes in `APIAccess` can be reverted and everything else should work now that user's own profile returns ruleset statistics

**Review by peppy — APPROVED** [review]
